components/tikv_util/src/resizable_threadpool.rs

```diff
@@ -1,161 +1,359 @@
+    where
+        Fut: Future<Output = ()> + Send + 'static,
+        {
+            let rc_runtime = match self.inner.upgrade() {
+                Some(runtime) => runtime,
+                None => return,
+            };
+        
+            let inner = match rc_runtime.inner.as_ref() {
+                Some(inner) => inner,
+                None => return,
+            };
+        
+            let handle = inner.handle().clone();
+            let task_count = rc_runtime.task_count.clone();
+        
+            task_count.fetch_add(1, Ordering::SeqCst);
+            
+            handle.spawn(async move {
+                fut.await;
+                task_count.fetch_sub(1, Ordering::SeqCst);
+            });
+        }
+
+    pub fn block_on<Fut>(&self, fut: Fut) -> Option<Fut::Output>
+    where
+        Fut: Future,
+    {
+        let (handle, task_count) = {
+            let rc_runtime = self.inner.upgrade()?;
+            let inner = rc_runtime.inner.as_ref()?;
+            let handle = inner.handle().clone();
+            let task_count = rc_runtime.task_count.clone();
+            (handle, task_count)
+        };    
+    
+        task_count.fetch_add(1, Ordering::SeqCst);
+    
+        Some(handle.block_on(async move {
+            let output = fut.await;
+            task_count.fetch_sub(1, Ordering::SeqCst);
+            output
+        }))
+    }
+}
+
 pub struct ResizableRuntime {
     pub size: usize,
+    count: usize,
     thread_name: String,
-    pool: Option<Arc<DaemonRuntime>>,
+    current_runtime: Arc<DeamonRuntime>,
+    used_runtime: Arc<Mutex<Vec<Arc<DeamonRuntime>>>>,
+    keeper: Option<Runtime>,
     replace_pool_rule: Box<dyn Fn(usize, &str) -> TokioResult<Runtime> + Send + Sync>,
     after_adjust: Box<dyn Fn(usize) + Send + Sync>,
 }
 
 impl ResizableRuntime {
     pub fn new(
+        thread_size: usize,
         thread_name: &str,
         replace_pool_rule: Box<dyn Fn(usize, &str) -> TokioResult<Runtime> + Send + Sync>,
         after_adjust: Box<dyn Fn(usize) + Send + Sync>,
     ) -> Self {
-        ResizableRuntime {
+        let init_thread_name = format!("{}-v0-{}", thread_name, thread_size);
+        let current_runtime = replace_pool_rule(thread_size, init_thread_name.as_str()).expect(
+            "failed to create tokio runtime for backup worker.");
+        let keeper = Builder::new_multi_thread()
+            .worker_threads(1)
+            .thread_name("rtkp")
+            .enable_all()
+            .build()
+            .expect("Failed to create runtime-keeper");
+
+        let mut ret = ResizableRuntime {
             size: 0,
+            count: 0,
             thread_name: thread_name.to_owned(),
-            pool: None,
+            current_runtime: Arc::new(DeamonRuntime {
+                inner: Some(current_runtime),
+                task_count: Arc::new(AtomicUsize::new(0)),
+            }),
+            used_runtime: Arc::new(Mutex::new(Vec::new())),
```

Comment:
I'm thinking if you can merge `used_runtime` and `current_runtime` into a single field, where the `current_runtime` is simply `used_runtime.last().unwrap()`. At that point the type may also be simplified to `Arc<Mutex<Vec<DeamonRuntime>>>` killing the double-Arc.

The tricky thing is to tell that GC loop not to delete the current runtime. Perhaps an additional `active: bool` field in DeamonRuntime to ensure `.retain()` not deleting it even when `task_count` is zero.

Reply:
We can't, becasue we need to create the Weak ref of DeamonRuntime as handle, this change only make things more complex.